### PR TITLE
refactor gc to only need to allocate virtual addresses as needed

### LINKED ISSUE #17987: command line flag to limit heap memory usage?
Should there be a command line flag to limit heap memory used by a Julia process?
Similar to `-Xms` and `-Xmx` for the JVM.

This may be useful while running Julia inside containers (docker) with memory limits or 
running multiple Julia processes on the same machine.

### PR DESCRIPTION
**Pull request by vtjnash**
From local testing, this doesn't seem to have any performance impact. To get virtual address space to line up with physical allocation a bit more would require also adjusting openblas to avoid allocation of it's 512 MB buffer (32 MB * 16 threads) during initialization.

fix #17987
alterate fix for #10390

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Could you explain what region0, region1, and region2 mean?

File: src/gc-pages.c [hunk 1]
```diff
@@ -96,113 +84,184 @@ static void jl_gc_alloc_region(region_t *region)
...
+            info.region0 = (*pregion0 = (region0_t*)calloc(1, sizeof(region0_t)));
+
+        i = REGION0_INDEX(ptr);
+        info.region0_i = i % 32;
+        info.region0_i32 = i / 32;
+        msk = (1 << info.region0_i);
+        info.region0->freemap[info.region0_i32] |= msk; // is free
+        pmeta = &info.region0->meta[i];
+        info.meta = (*pmeta = &page_meta[pg]);
+    }
```
Comment (yuyichao) [inline src/gc-pages.c]: The allocation here can probably be cleaned up later.....  (we can have a version of perm_alloc that guarantees zero filling).

File: src/gc.h [hunk 2]
```diff
@@ -232,31 +271,49 @@ STATIC_INLINE void *gc_ptr_clear_tag(void *v, uintptr_t mask)
...
-        char *end = begin + region->pg_cnt * sizeof(jl_gc_page_t);
-        if ((char*)ptr >= begin && (char*)ptr <= end) {
-            return region;
-        }
-    }
-    return NULL;
+    uintptr_t data = ((uintptr_t)_data);
+    unsigned i;
+    i = REGION_INDEX(data);
+    region1_t *r1 = memory_map.meta1[i];
```
Comment (yuyichao) [inline src/gc.h]: This can probably use a compile time optimization too. (Though I doubt 32bits matters too much.......)

**Review by yuyichao — APPROVED** [review]
LGTM in general. The `page_metadata` fast path can be tweaked a little bit. Other slow paths can probably be tweaked later.

**Comment by yuyichao** [comment]
I believe the "regions" forms a 2-3 level "page table".

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

File: src/gc-pages.c [hunk 3]
```diff
@@ -96,113 +84,184 @@ static void jl_gc_alloc_region(region_t *region)
...
+    info.region0->freemap[info.region0_i32] &= ~(uint32_t)(1 << info.region0_i);
+    info.region0->allocmap[info.region0_i32] |= (uint32_t)(1 << info.region0_i);
+    info.region1->allocmap0[info.region1_i32] |= (uint32_t)(1 << info.region1_i);
+    memory_map.allocmap1[info.region_i32] |= (uint32_t)(1 << info.region_i);
 
-    region->allocmap[i] |= (uint32_t)(1 << j);
-    void *ptr = region->pages[i * 32 + j].data;
 #ifdef _OS_WINDOWS_
-    VirtualAlloc(ptr, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
+    VirtualAlloc(meta->data, GC_PAGE_SZ, MEM_COMMIT, PAGE_READWRITE);
```
Comment (ararslan) [inline src/gc-pages.c]: It looks like this is using `meta` outside of its scope (defined on line 178 inside a block) which is causing the Windows failures.

File: src/gc-pages.c [hunk 4]
```diff
@@ -96,113 +84,184 @@ static void jl_gc_alloc_region(region_t *region)
             jl_throw(jl_memory_exception);
```
Comment (ararslan) [inline src/gc-pages.c]: This is getting hit on 32-bit Linux

File: src/gc.c [hunk 5]
```diff
@@ -590,7 +590,7 @@ STATIC_INLINE void gc_setmark_pool_(jl_ptls_t ptls, jl_taggedvalue_t *o,
 STATIC_INLINE void gc_setmark_pool(jl_ptls_t ptls, jl_taggedvalue_t *o,
                                    uint8_t mark_mode)
 {
-    gc_setmark_pool_(ptls, o, mark_mode, page_metadata(o));
+    gc_setmark_pool_(ptls, o, mark_mode, page_metadata_ext(o).meta);
```
Comment (vtjnash) [inline src/gc.c]: yeah, main difference is that `_ext` requires it to be a valid page, while `page_metadata` did not

Comment (vtjnash) [inline src/gc.c]: also, `_ext` is the fast path

File: src/gc-pages.c [hunk 6]
```diff
@@ -9,32 +9,21 @@
 extern "C" {
 #endif
 
-// A region is contiguous storage for up to DEFAULT_REGION_PG_COUNT naturally aligned GC_PAGE_SZ pages
```
Comment (tkelman) [inline src/gc-pages.c]: don't just delete this, update it if no longer true

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21135#issuecomment-288535031) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d3317653960a9a1d39f40f481978ed89a40d6d15/d07bb51_vs_89609cf/report.md). cc @jrevels

**Comment by tkelman** [comment]
the difference between pagetable, pagetable0, and pagetable1 is still not described very clearly

**Comment by vtjnash** [comment]
There is no difference, just level h=0, h=1, and h=2 of the table. C is just really encouraging of copy-pasting.

**Comment by tkelman** [comment]
these look copy pasted but in several cases aren't quite, the 0 function refers to 1, or similar. is this intentional? if so, it should be explained.

**Comment by StefanKarpinski** [comment]
@vtjnash: I don't really understand why you can't just say what they are for. If they're the same, why are there three of them? Why not just one?

**Comment by vtjnash** [comment]
> Why not just one?

Exactly. This is why people don't like C. Three data-structures that differ only by their distance from the root of the tree each have their own independently unique type.

> these look copy pasted but in several cases aren't quite

Of course, it's a table of height 3, so level 2 points at 1 and level 2 points at level 1 which contains the leaf nodes. That means I can hand-optimize away some of the conditional checks at each level, so that the copies of the code ends up littered with special cases.

**Comment by tkelman** [comment]
> Of course, it's a table of height 3, so level 2 points at 1 and level 2 points at level 1 which contains the leaf nodes.

Then describe that in comments, it's not obvious without any explanation.

**Comment by tkelman** [comment]
`gc_clear_mark_page*`, `gc_verify_tags_pagetable*`, `gc_stats_pagetable*`, `gc_count_pool_pagetable*`, `sweep_pool_page*` still have no descriptive comments

**Comment by vtjnash** [comment]
We don't usually have descriptive comments on helper functions (unless the work they do is non-intuitive). In this case, I've just split out the existing iterators into their own functions.

**Comment by tkelman** [comment]
These aren't intuitive - the problem isn't that they're repetitive, it's that they look copy pasted but are each slightly different in ways that aren't at all obvious if you didn't write the code. For maintainability of the code base, please explain the scheme better. We need many many more comments throughout, and confusing aspects are the best place to fix that first.

File: src/gc.h [hunk 7]
```diff
@@ -151,13 +151,20 @@ typedef struct {
...
-//   and can use that to partially unroll and optimize various conditions at each level.
-//  These constants define the branching factor.
-//  These constants plus GC_PAGE_LG2 therefore must sum to sizeof(void*).
-//  They should all be multiples of 32 (sizeof(uint32_t)), except REGION2_PG_COUNT is also allowed to be 1.
+//  The complete address space is divided up into a multi-level page table.
+//  The three levels have similar but slightly different structures:
+//    - pagetable0_t: the top/root level
+//    - pagetable1_t: the middle level
+//    - pagetable2_t: the bottom/leaf level
+//  Corresponding to these simlar structures is a large amount of repetetive
```
Comment (ararslan) [inline src/gc.h]: repetetive -> repetitive (here and below)

File: src/gc.h [hunk 8]
```diff
@@ -151,13 +151,20 @@ typedef struct {
...
+//  code that is nearly the same but not identical. It could be made less
+//  repetetive with C macros, but only at the cost of debuggability. The specialized
+//  structure of this representation allows us to partially unroll and optimize
+//  various conditions at each level.
+
+//  The following constants define the branching factors at each level.
+//  They constants and GC_PAGE_LG2 must therefore sum to sizeof(void*).
```
Comment (ararslan) [inline src/gc.h]: they -> the

**Comment by tkelman** [comment]
the gc.h comment helps, but I meant inline on the functions I referred to - the pieces that differ, when a function called 0 refers to a variable called 1, or a variable called 1 has a field named 0, etc, is what's confusing and could use point-of-use indications of intent

File: src/gc.h [hunk 9]
```diff
@@ -147,33 +143,90 @@ typedef struct {
...
+//    Tag: sizeof(jl_taggedvalue_t)
+//    Data: <= osize - sizeof(jl_taggedvalue_t)
+
+// Memory map:
+//  The complete address space is divided up into a multi-level page table.
+//  The three levels have similar but slightly different structures:
+//    - pagetable0_t: the bottom/leaf level (covers the contiguous addresses)
+//    - pagetable1_t: the middle level
+//    - pagetable2_t: the top/leaf level (covers the entire virtual address space)
+//  Corresponding to these simlar structures is a large amount of repetitive
```
Comment (tkelman) [inline src/gc.h]: similar

gc_clear_mark_*, gc_verify_tags_*, gc_stats_pagetable*, and gc_count_pool_pagetable* still look misleading and could use comments

Comment (vtjnash) [inline src/gc.h]: Those are debugging functions not actually part of he gc. They are usually commented out, so I wouldn't worry about them.

Comment (tkelman) [inline src/gc.h]: of course you wouldn't, you wrote them

**Comment by tkelman** [comment]
Either this or #21308 broke the power build https://build.julialang.org/builders/package_tarballppc64le/builds/133

**Pull request merged**